File: binding/golang/service/ip2region.go
// Copyright 2022 The Ip2Region Authors. All rights reserved.
// Use of this source code is governed by a Apache2.0-style
// license that can be found in the LICENSE file.

package service

import (
	"fmt"
	"time"

	"github.com/lionsoul2014/ip2region/binding/golang/xdb"
)

// ---
// Ip2Region service
// 1. Unified query interface to IPv4 and IPv6 address.
// 2. Concurrency search support.
//
// @Author Lion <chenxin619315@gmail.com>
// @Date   2025/12/05

type Ip2Region struct {
	// v4 pool for cache policy vIndex or NoCache
	v4Pool *SearcherPool

	// v4 xdb searcher for full in-memeory search
	v4InMemSearcher *xdb.Searcher

	// v6 pool for cache policy vIndex or NoCache:w
	v6Pool *SearcherPool

	// v6 xdb searcher for full in-memeory search
	v6InMemSearcher *xdb.Searcher
}

// create a new Ip2Region service with specified v4 and v6 config.
// set it to nil to disabled the specified search for the specified version.
func NewIp2Region(v4Config *Config, v6Config *Config) (*Ip2Region, error) {
	var err error

	// check and init the v4 pool or in-memory searcher
	var v4Pool *SearcherPool
	var v4InMemSearcher *xdb.Searcher
	if v4Config == nil {
		// with IPv4 disabled ?
		v4Pool = nil
		v4InMemSearcher = nil
	} else if v4Config.cachePolicy == BufferCache {
		v4Pool = nil
		v4InMemSearcher, err = xdb.NewWithBuffer(v4Config.ipVersion, v4Config.cBuffer)
		if err != nil {
			return nil, fmt.Errorf("failed to create v4 in-memory searcher: %w", err)
		}
	} else {
		v4InMemSearcher = nil
		v4Pool, err = NewSearcherPool(v4Config)
		if err != nil {
			return nil, fmt.Errorf("failed to create v4 searcher pool: %w", err)
		}
	}

	// check and init the v6 pool or in-memory searcher
	var v6Pool *SearcherPool
	var v6InMemSearcher *xdb.Searcher
	if v6Config == nil {
		v6Pool = nil
		v6InMemSearcher = nil
	} else if v6Config.cachePolicy == BufferCache {
		v6Pool = nil
		v6InMemSearcher, err = xdb.NewWithBuffer(v6Config.ipVersion, v6Config.cBuffer)
		if err != nil {
			return nil, fmt.Errorf("failed to create v6 in-memory searcher: %w", err)
		}
	} else {
		v6InMemSearcher = nil
		v6Pool, err = NewSearcherPool(v6Config)
		if err != nil {
			return nil, fmt.Errorf("failed to create v6 in-memeory searcher pool: %w", err)
		}
	}

	return &Ip2Region{
		v4Pool:          v4Pool,
		v4InMemSearcher: v4InMemSearcher,

		v6Pool:          v6Pool,
		v6InMemSearcher: v6InMemSearcher,
	}, nil
}

// create the ip2region search service with the specified v4 & v6 xdb path.
// with default cache policy VIndexCache and default searchers = 20
func NewIp2RegionWithPath(v4XdbPath string, v6XdbPath string) (*Ip2Region, error) {
	var err error

	// create v4 config with default config items
	var v4Config *Config
	if v4XdbPath == "" {
		v4Config = nil
	} else {
		v4Config, err = NewV4Config(VIndexCache, v4XdbPath, 20)
		if err != nil {
			return nil, fmt.Errorf("failed to create v4 config: %w", err)
		}
	}

	// create v6 config with default config items
	var v6Config *Config
	if v6XdbPath == "" {
		v6Config = nil
	} else {
		v6Config, err = NewV6Config(VIndexCache, v6XdbPath, 20)
		if err != nil {
			return nil, fmt.Errorf("failed to create v6 config: %w", err)
		}
	}

	return NewIp2Region(v4Config, v6Config)
}

func (ip2r *Ip2Region) SearchByStr(ipStr string) (string, error) {
	ipBytes, err := xdb.ParseIP(ipStr)
	if err != nil {
		return "", err
	}

	return ip2r.Search(ipBytes)
}

func (ip2r *Ip2Region) Search(ipBytes []byte) (string, error) {
	if l := len(ipBytes); l == 4 {
		return ip2r.v4Search(ipBytes)
	} else if l == 16 {
		return ip2r.v6Search(ipBytes)
	} else {
		return "", fmt.Errorf("invalid byte ip address with len=%d", l)
	}
}

func (ip2r *Ip2Region) v4Search(ipBytes []byte) (string, error) {
	if ip2r.v4InMemSearcher != nil {
		return ip2r.v4InMemSearcher.Search(ipBytes)
	}

	// v4 search is disabled
	if ip2r.v4Pool == nil {
		return "", nil
	}

	v4Searcher := ip2r.v4Pool.BorrowSearcher()
	defer ip2r.v4Pool.ReturnSearcher(v4Searcher)
	return v4Searcher.Search(ipBytes)
}

func (ip2r *Ip2Region) v6Search(ipBytes []byte) (string, error) {
	if ip2r.v6InMemSearcher != nil {
		return ip2r.v6InMemSearcher.Search(ipBytes)
	}

	// v6 search is disabled
	if ip2r.v6Pool == nil {
		return "", nil
	}

	v6Searcher := ip2r.v6Pool.BorrowSearcher()
	defer ip2r.v6Pool.ReturnSearcher(v6Searcher)
	return v6Searcher.Search(ipBytes)
}

func (ip2r *Ip2Region) Close() {
	ip2r.CloseTimeout(time.Second * 10)
}

func (ip2r *Ip2Region) CloseTimeout(d time.Duration) {
	if ip2r.v4Pool != nil {
		ip2r.v4Pool.CloseTimeout(d)
	}

	if ip2r.v6Pool != nil {
		ip2r.v6Pool.CloseTimeout(d)
	}
}

File: binding/java/src/main/java/org/lionsoul/ip2region/service/Ip2Region.java
// Copyright 2022 The Ip2Region Authors. All rights reserved.
// Use of this source code is governed by a Apache2.0-style
// license that can be found in the LICENSE file.

package org.lionsoul.ip2region.service;

import java.io.IOException;

import org.lionsoul.ip2region.xdb.InetAddressException;
import org.lionsoul.ip2region.xdb.Searcher;
import org.lionsoul.ip2region.xdb.Util;
import org.lionsoul.ip2region.xdb.XdbException;

/**
 * ip2region searcher manager service to provider:
 * 1. Unified query interface for IPv4 and IPv6 address.
 * 2. Concurrency search support.
 * 
 * @Author Lion <chenxin619315@gmail.com>
 * Date    2025/11/21
*/
public class Ip2Region {

    /* v4 pool for cache policy vIndex or NoCache */
    private final SearcherPool v4Pool;

    /* v4 xdb searcher for cache policy cBuffer */
    private final Searcher v4InMemSearcher;

    /* v6 pool for cache policy vIndex or NoCache*/
    private final SearcherPool v6Pool;

    /* v6 xdb searcher for cache policy cBuffer */
    private final Searcher v6InMemSearcher;

    public static final Ip2Region create(final Config v4Config, final Config v6Config) throws IOException {
        return new Ip2Region(v4Config, v6Config).init();
    }

    public static final Ip2Region create(final String v4XdbPath, final String v6XdbPath) throws IOException, XdbException {
        return new Ip2Region(v4XdbPath, v6XdbPath).init();
    }

    /**
     * init the ip2reigon with two xdb file path and default cachePolicy vIndex.
     * set it to null to disabled the search for specified version
     * 
     * @param v4XdbPath
     * @param v6XdbPath
     * @throws XdbException 
     * @throws IOException 
    */
    protected Ip2Region(String v4XdbPath, String v6XdbPath) throws IOException, XdbException {
        this(
            v4XdbPath == null ? null : Config.custom().setXdbPath(v4XdbPath).asV4(), 
            v6XdbPath == null ? null : Config.custom().setXdbPath(v6XdbPath).asV6()
        );
    }

    /**
     * init the ip2region with specified config.
     * set it to null for disabled the search for specified version
     * 
     * @param v4Config
     * @param v6Config
     * @throws IOException
    */ 
    protected Ip2Region(Config v4Config, Config v6Config) throws IOException {
        if (v4Config == null) {
            // @Note: with IPv4 disabled ?
            this.v4InMemSearcher = null;
            this.v4Pool = null;
        } else if (v4Config.cachePolicy == Config.BufferCache) {
            this.v4InMemSearcher = Searcher.newWithBuffer(v4Config.ipVersion, v4Config.cBuffer);
            this.v4Pool = null;
        } else {
            this.v4InMemSearcher = null;
            this.v4Pool = new SearcherPool(v4Config);
        }

        if (v6Config == null) {
            // @Note: with IPv6 disabled ?
            this.v6InMemSearcher = null;
            this.v6Pool = null;
        } else if (v6Config.cachePolicy == Config.BufferCache) {
            this.v6InMemSearcher = Searcher.newWithBuffer(v6Config.ipVersion, v6Config.cBuffer);
            this.v6Pool = null;
        } else {
            this.v6InMemSearcher = null;
            this.v6Pool = new SearcherPool(v6Config);
        }
    }

    // init the current ip2region service
    protected Ip2Region init() throws IOException {
        if (v4Pool != null) {
            v4Pool.init();
        }

        if (v6Pool != null) {
            v6Pool.init();
        }

        return this;
    }

    public String search(String ipString) throws InetAddressException, IOException, InterruptedException {
        return search(Util.parseIP(ipString));
    }

    public String search(byte[] ipBytes) throws InetAddressException, IOException, InterruptedException {
        if (ipBytes.length == 4) {
            return v4Search(ipBytes);
        } else if (ipBytes.length == 16) {
            return v6Search(ipBytes);
        } else {
            throw new InetAddressException("invalid byte ip address with length=" + ipBytes.length);
        }
    }

    protected String v4Search(final byte[] ipBytes) throws IOException, InetAddressException, InterruptedException {
        if (v4InMemSearcher != null) {
            return v4InMemSearcher.search(ipBytes);
        }

        // IPv4 search is disabled
        if (v4Pool == null) {
            return null;
        }

        final Searcher searcher = v4Pool.borrowSearcher();
        try {
            return searcher.search(ipBytes);
        } finally {
            v4Pool.returnSearcher(searcher);
        }
    }

    protected String v6Search(final byte[] ipBytes) throws IOException, InetAddressException, InterruptedException {
        if (v6InMemSearcher != null) {
            return v6InMemSearcher.search(ipBytes);
        }

        // IPv6 search is disabled
        if (v6Pool == null) {
            return null;
        }

        final Searcher searcher = v6Pool.borrowSearcher();
        try {
            return searcher.search(ipBytes);
        } finally {
            v6Pool.returnSearcher(searcher);
        }
    }

    public void close() throws InterruptedException {
        close(10000);
    }

    public void close(long timeoutMillis) throws InterruptedException {
        if (v4Pool != null) {
            v4Pool.close(timeoutMillis);
        }

        if (v6Pool != null) {
            v6Pool.close(timeoutMillis);
        }
    }

}